Add page on Linux shell (#91)
se-edu/learningresources: contents/linux/shell.md

<frontmatter>
  title: Introduction to Linux Shell
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

{{ navbar | safe }}

<div class="website-content">

> # Introduction to Linux bash shell 

Reviewer: Could I propose that this article be called `Introduction to Bash shell`? Similar reasons as the first comment.

Author: [Wang Junming](https://github.com/junming403)

## Overview

> This chapter assumes that the reader is familiar with general concepts of operating systems. If you are not familiar with basic operating system concepts, I recommend you pick up the basic concepts first. The classic textbook of operating system **Mordern Operating Systems by Tanenbaum** is highly recommended. 

Reviewer 2: Do we really need that much familiarity with Operating Systems to understand how to use the basic bash shell?

The textbook is used for introductory courses such as Introduction to Operating Systems. Seems a bit much here to read that before coming here?

Why not you just link a simple article or something?

Reviewer 3: Agree. This article isn't even for teaching how to use the bash shell - it's why one should use the bash shell.  I think that if the person is still at that stage, knowledge at this level should not be required.

Author: While in my understanding, one can only make good use of bash shell only if he understand how  OS kernel works. Indeed I agree that without learning basic concepts of OS, one can still use OS, but it's always good to understand how tools works before using the tool? I will mention this Book is a not prerequisite to read this article instead.

<img src="images/shell-kernel.png" alt="shell-kernel" style="width:400px;"/>

> Computer can be divided into 3 abstraction layers, hardware, kernel and applications. Users cannot control the hardware directly, instructions have to be given through kernel as kernel is the one that controls the hardware. 

Reviewer 3: Is this explanation necessary for understanding the article? It feels too complex for the target audience, using terms that may be unfamiliar to them - abstraction, kernel etc.

Author: I think it's essential, because I want my audience to understand how shell works before start to use it, as such one can really master it.

> Shell is the interface through which we can pass instructions to kernel, and then it could be executed on the hardware. In this chapter, we will look into **Linux command-line based shell(bash)**.

Reviewer: `bash` is not the only shell on Linux; see first comment.

> > Linux is the best-known and most-used open source operating system. As an operating system, Linux is software that sits underneath all of the other software on a computer, receiving requests from those programs and relaying these requests to the computer’s hardware.

Reviewer: `best-known and most-used`: a reference to back this up? A causal reader would think this is Windows.

[**bash**](https://www.gnu.org/software/bash/manual/html_node/) stands for **Bourne Again SHell**, an enhanced version of the original Unix Shell program. It has become the standard shell of various linux distributions. 

> ## Why use command-line based shell?

Reviewer 2: This section seems very technical! You could also add some points to convince total beginners to switch over to bash? People who have zero experience with command line may not be able to follow along with the code snippets you have put up.

Author: Actually I do assume some understanding of OS in order to understand this article, because OS is an essentially a must-learn for everyone of CS background.  I have listed points below for why should get started to using shell.
1. `Remote Management` 
2. `Stability`
3. `Freedom`
4. `Simplicity`

> A common question many people would ask is why would we learn command-line based shell as it seems to be very hard to use. There are tons of GUI tools out there, they seems to look more user-friendly and powerful. From my perspective, there are four major reasons for starting to use command line.

Reviewer 3: Expand it out, it's the first time you're using the phrase.
```suggestion
A common question many people would ask is why would we learn command-line based shell as it seems to be very hard to use. There are tons of Graphical User Interface (GUI) tools out there, they seems to look more user-friendly and powerful. From my perspective, there are four major reasons for starting to use command line.
```

Reviewer 3: More thoughts 
```suggestion
A common question many people ask is why should we learn command-line based shell as it seems to be very hard to use. There are tons of Graphical User Interface (GUI) tools out there, that appear to be more user-friendly and powerful. From my perspective, there are four major reasons to use the command line:
```

Author: Thanks, that looks better.

> - **Remote Management**: Use over remote command-line shell makes it much easier to perform tasks remotely, such as when a developer is connecting from home. A remote command-line shell usually requires less bandwidth.

Reviewer 3: If I'm not familiar with remote shells, I could argue that the solutions that I'm using work too - say, teamviewer to perform remote control. Arguably easier with a GUI too. The only point here that beats it is less bandwidth. Thus, it feels pretty weak - perhaps mention it's better supported with many other ssh programs, require less cpu usage for the connection, etc?

Reviewer: I would say that a major reason to use a shell for remote management is _lag_: doing things via the shell is so much faster.

Author: > I would say that a major reason to use a shell for remote management is _lag_: doing things via the shell is so much faster.

Exactly, like what I have mentioned, "A remote command-line shell usually requires less bandwidth.", I will emphasise this more.

- **Stability**: No matter what linux distribution you are running, the command line shell is always available and usage is almost the same. However, GUI tools might be different in different distributions, make it very hard to learn.

> - **Freedom**: You will have complete control in command-line shell. GUIs usually tends to simplifying things, giving user fewer options to make things easier. In command line there is no such simplifications, you will have complete control of what you are doing.

Reviewer 3: Rephrase + grammer
```suggestion
- **Freedom**: You will have complete control in the command-line shell. GUIs tends to simplify things, giving the user fewer options. However, the command line always provides the full suite of options, giving you complete control.```

> - **Simplicity**: Although you might find command-line hard to use at first, but after using it enough you'll eventually come to the point where the command-line is far easier to use than trying to find options in the drop-down menus and alike of a GUI. 

Reviewer 3: Not sure if better.  Just a thought. The original feels like run on sentence though.
```suggestion
- **Simplicity**: While you might find command-line difficult to grasp, eventual usage will familiarize yourself such that the command-line is far easier to use than trying to find options in the drop-down menus and alike of a GUI. 
```

### Benefit: Program with bash script

> One important feature in bash is **bash script**. It enables us to do `task automation`. Suppose you are a system admin, there is a task you need to perform daily without any changes. Instead of typing in the set of commands for the task everyday, you can store these commands in a file and run the script.

Reviewer 3: Perhaps
```suggestion
One important feature in bash is the **bash script**. It enables us to do `task automation`. For example, suppose that you are a system admin and there is a task you need to perform daily. Instead of typing in the commands for the task daily, you can store these commands in the **bash script** and run it.
```

Reviewer: * Instead of the abstract example of `a task you need to perform daily`, it would be clearer to give a concrete example, say, daily rotation of log files.

* Why is `task automation` in `code` styling? Emphasis should probably be done with _italics_ or **bolding**.

> Here's one example of bash script, suppose you have to check if some given file exist periodically. The following script will ask for your input and output the results. `filename` is a variable used to store user input, `filetype` and `perm` stores corresponding attributes of the file for output.

Reviewer 3: ```suggestion
For example, suppose you have to check if some file exists periodically. The following script will ask for your input and output the results. `filename` is a variable used to store user input, `filetype` and `perm` stores corresponding attributes of the file for output.
```

```bash
#!/bin/bash
 
echo -e "Please input a filename, I will check the filename's type and permission.\n\n"

# waiting for user input.
read -p "Input a filename : "  filename

# check if file exists.

> test -z $filename && echo "You MUST input a filename." && exit 0

Reviewer: The Google Shell Style Guide [recommends `if [[...]]` for tests](https://google.github.io/styleguide/shell.xml?showone=Test,_%5B_and_%5B%5B#Test,_%5B_and_%5B%5B).

test ! -e $filename && echo "The filename '$filename' DO NOT EXIST" && exit 0

# check file type.
test -f $filename && filetype="regular file"
test -d $filename && filetype="directory"

# check file permission.
test -r $filename && perm="readable"
test -w $filename && perm="$perm writable"
test -x $filename && perm="$perm executable"

# output result.
echo "The filename: $filename is a $filetype"
echo "And permissions are : $perm"
```

> The above bash script is still just a collection of bash commands. Although it saves us a lot of works if it is a daily routine task, it still can be done by typing those commands line by line. What makes bash script really exciting is the semantics of `conditionals`, `loops` and `functions`. Which makes bash script [**turning complete**](https://en.wikipedia.org/wiki/Turing_completeness).

Reviewer 3: Grammar. Maybe bold bash script?
```suggestion
The above bash script is still just a collection of bash commands. Although it saves us a lot of work if it is a routine task, it can still be done by typing those commands line by line. What makes bash script really exciting is the semantics of `conditionals`, `loops` and `functions`. Which makes bash script [**turning complete**](https://en.wikipedia.org/wiki/Turing_completeness).
```

Reviewer: Why do we care that Bash scripts are Turing complete? I feel that this is a theoretical point that may not be of practical relevance; the above example is good enough to demonstrate its power.

> The following is an example of bash script that make use of conditionals and loops. Suppose you want to let user input a directory name, check if it exists, and then output permissions for all files inside that directory.

Reviewer 3: Grammar
```suggestion
The following is another example of bash script that utilizes conditionals and loops. Suppose you want to allow the user to input a directory name, check if it exists and output permissions for all files inside that directory.
```

```bash
#!bin/bash

# read user input and check if directory exists.
read -p "Please enter a directory name: " dir
if [ "$dir" == "" -o ! -d "$dir" ]; then
	echo "The $dir is NOT exist in your system"
	exit 1
fi

# output permissions for each of the file under the directory.
filelist=$(ls $dir)
for filename in $filelist
do 
	perm=""
	test -r "$dir/$filename" && perm="readable"
	test -w "$dir/$filename" && perm="$perm writable"
	test -x "$dir/$filename" && perm="$perm executable"
	echo "The file $dir/$filename 's permission is $perm" 
done
```
**Functions** in bash script just like functions in normal programming languages, more semantics of bash script can be found [here](https://www.gnu.org/software/bash/manual/html_node/). 

> bash script allows us to automate the frequently performed operations, it is easy to use and can be executed on any Unix-like operating systems without any modifications.

Reviewer 3: Perhaps something like "In conclusion," or "In summary,"

### Benefit: Data Stream Redirection

> In linux, everything is abstracted as file, including **input**, **output**, and **error** streams. **Redirection** is a feature in Linux such that when executing a command/user program, you can change the standard input/output devices by changing its [file descriptor](https://en.wikipedia.org/wiki/File_descriptor). Each command and user program in linux is an application with input, output and errors. We can redirect the output of one application to the input of another, as such, we combined the two applications together as if we build a pipe between them.

Reviewer 3: Grammar + suggested change near the end.
```suggestion
In linux, everything is abstracted as a file, including **input**, **output**, and **error** streams. **Redirection** is a feature in Linux such that when executing a command/user program, you can change the standard input/output devices by changing its [file descriptor](https://en.wikipedia.org/wiki/File_descriptor). Each command and user program in linux is an application with input, output and error streams. A common use case is to redirect the output of one application to the input of another, combining the two applications together as if there was pipe between them.
```

![file-descriptor](images/file-descriptor-illustration.jpg)

> For example, suppose you just finished your coding assignment, the resulting executable `calculate` takes a user input, do the calculation, and gives the output. To run it, you simply types:

Reviewer 3: Interesting, here you use "for example" but above it was "The following is an example of..." 
:P 
Would replace do with "performs"

```bash
./calculate
```

> But you are not satisfied with manual testing, instead, you want to test with larger data set file `DataBundle`. And verify the output with another executable `verify`, which takes in the result as input and verify its correctness and output **PASS** or **FAIL**. Here's how we can do it.

Reviewer 3: ```suggestion
But you are not satisfied with manual testing, instead, you want to test with a larger data set file `DataBundle` and verify the output with another executable `verify`. `verify` takes in the result as input and verifies its correctness. It outputs **PASS** or **FAIL** depending on if it passes or fails. Here's how we can do it.
```

```bash
./calculate < DataBundle > result
./verify < result
```
Notice in this case, `calculate` uses the data in `DataBundle` as input, and output the results to the file `result`. Then, `result` is taken as input to `verify`, and the final verification result is printed on screen.

> Furthermore, if the code we write is buggy, it throws exception and error messages are directed to error output, then our result file will be empty! To handle it properly, we can also redirect error output to a file `errors`, and verify it is empty(such that no error occurs during execution) before we execute `verify`.

Reviewer 3: ```suggestion
However, if the code we write is buggy, any thrown exception and error messages are directed to error output, resulting in an empty result file! To handle this scenario, we can also redirect error output to a file `errors`, and verify it is empty (meaning that no error occurred during execution) before we execute `verify`.
```

```bash
#!/bin/bash
./calculate < DataBundle > result 2> errors

> if [ -s errors]

Reviewer: `if [ -s errors ]`?

then
     ./verify < result
else
     echo "error occured during execution."
fi
```

> Another useful command is `|`, which chained linux commands together, as such the output of previous commands is passed as input to the next command.

Reviewer 4: one nice thing to mention in here could be `tee`, and since you used the different streams here, might wanna put some link to it with more explanation?

> Suppose you want to check the status of process `p` running on your system. Thus you typed in `ps aux`, and all the processes are listed, but there are too many processes that you can't find `p`. As such, you can pass the result of `ps aux` to `grep 'p'` to capture process `p`'s status. Thus the command will be look like `ps aux | grep 'p'`.

Reviewer 3: You should mention briefly what ps and grep does.

Reviewer: I suggest using a concrete example; perhaps checking the status of long-running process (eg. zipping a large file with `gzip`).

> Input/output streams can easily be redirected, allowing information to be sent or received from files or other applications(commands and user programs). This can mean test data can be easily supplied or output captured.

Reviewer 4: can mention list of common applications used, `awk`, `sed`, etc. so that they know what to start looking for?

> Input/output streams can easily be redirected, allowing information to be sent or received from files or other applications(commands and user programs). This can mean test data can be easily supplied or output captured.

Reviewer 3: ```suggestion
Input/output streams can be easily redirected, allowing information to be sent or received from files or other applications (commands and user programs). This means that test data can be easily supplied or output captured to a file.
```

For example, if you are new to bash and want to learn a new command `grep`, you can simply type in `man grep` in the terminal and a help page will show up. It thoroughly include **What**, **How** as well as many other details of the command.

```
GREP(1)                   BSD General Commands Manual                  GREP(1)

NAME
     grep, egrep, fgrep, zgrep, zegrep, zfgrep -- file pattern searcher

SYNOPSIS
     grep [-abcdDEFGHhIiJLlmnOopqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
          [-e pattern] [-f file] [--binary-files=value] [--color[=when]]
          [--colour[=when]] [--context[=num]] [--label] [--line-buffered]
          [--null] [pattern] [file ...]

DESCRIPTION
     The grep utility searches any given input files, selecting lines that
     match one or more patterns.  By default, a pattern matches an input line
     if the regular expression (RE) in the pattern matches the input line
     without its trailing newline.  An empty expression matches every line.
     Each input line that matches at least one of the patterns is written to
     the standard output ......

```

> ## How to Get Started with linux command-line based shell?

Reviewer 2: This part here just feels like your saying, 'jump right in and start using it' !

I liked your git example here! Maybe you can elaborate a bit more on specific use cases when I should turn to the command line?

Author: > I originally thought that this article would be on learning how to use the linux shell, but it's really just why you should start learning.`

That's because the requirement of book chapter is to focus more on what and why.

> In my opinion, there is no need to learning linux command-line based shell in particular, because there are tons of linux commands out there, it is impossible to remember and master them all at one time without actual practices. Instead, when it comes to the time that you have to use it, try to use the terminal instead of GUI based tool. Although it might be hard at first, it will benefit you more later. For example, `git`, I highly recommend get started to use command-line based git instead of visualisation tools such as `Source Tree`.

Reviewer 3: Simplified some text
```suggestion
In my opinion, there is no need to learning linux command-line based shell in particular, because there are tons of linux commands out there, it is impossible to remember and master them all at once without actual practice. Instead, when it comes time to use it, try to use the terminal instead of GUI based tool. Although it might be hard at first, it will benefit you more later. For example, `git`, I highly recommend get started to use command-line based git instead of visualisation tools such as `Source Tree`.
```

> # Where to Go from Here?

Reviewer 3: Perhaps a link to start to a tutorial to start learning shell commands from the very basics? Like ls, cd, mkdir, mv, cp etc.

Reviewer: Some useful resources:

- The [Google Shell Style Guide](https://google.github.io/styleguide/shell.xml) is a good baseline for good shell scripting style.
- Gentoo has a good [`bash` shell reference](https://devmanual.gentoo.org/tools-reference/bash/).
- Apple's [Shell Scripting Primer](https://developer.apple.com/library/archive/documentation/OpenSource/Conceptual/ShellScripting/Introduction/Introduction.html) is a in-depth introduction to shell scripting for beginners.

- [Introduction to Linux](http://www.linfo.org/newbies.html) is a good introduction to Linux and contains many useful resource links.
- [Bash Reference Manual](https://www.gnu.org/software/bash/manual/html_node/) is a good reference manual of linux bash shell. 
- [Online man Page](https://www.linux.org/docs/) is a collection of  `man pages`  online.
- [linux.org](https://www.linux.org/) is the Linux organization home page, you can find all relevant information about linux here. 

> - One big advantage of linux is its open source feature, you can read [Linux Source Code](https://github.com/torvalds/linux) by your own interest. 

Reviewer: Someone who is just being introduced to Bash scripting may not get much out of Linux kernel code 🤔
